In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
ruta_archivo_zip = '/content/drive/MyDrive/abalone.zip'


In [ ]:
import zipfile

ruta_archivo_zip = '/content/drive/MyDrive/abalone.zip'

ruta_destino = '/content/drive/MyDrive/ruta/destino'


with zipfile.ZipFile(ruta_archivo_zip, 'r') as zip_ref:
    zip_ref.extractall(ruta_destino)


archivos_extraidos = zip_ref.namelist()
print("Archivos extraídos:", archivos_extraidos)


Archivos extraídos: ['Index', 'abalone.data', 'abalone.names']


In [ ]:
import zipfile
import pandas as pd
import matplotlib.pyplot as plt


ruta_archivo_zip = '/content/drive/MyDrive/abalone.zip'


archivo_datos = 'abalone.data'


with zipfile.ZipFile(ruta_archivo_zip, 'r') as zip_ref:
    with zip_ref.open(archivo_datos) as file:
        # Leemos los datos del archivo y decodificarlos
        datos = file.read().decode('utf-8')

# Dividimos los datos por saltos de línea y comas
filas = [fila.split(',') for fila in datos.strip().split('\n')]

# Creamos un DataFrame de pandas
df = pd.DataFrame(filas, columns=['Sexo', 'Longitud', 'Diámetro', 'Altura', 'Peso_entero', 'Peso_manzana', 'Peso_vísceras', 'Peso_concha', 'Anillos'])

# Mostramos el DataFrame
print(df)

# Graficamos la distribución de la longitud de los abalones
df.info()


     Sexo Longitud Diámetro Altura Peso_entero Peso_manzana Peso_vísceras  \
0       M    0.455    0.365  0.095       0.514       0.2245         0.101   
1       M     0.35    0.265   0.09      0.2255       0.0995        0.0485   
2       F     0.53     0.42  0.135       0.677       0.2565        0.1415   
3       M     0.44    0.365  0.125       0.516       0.2155         0.114   
4       I     0.33    0.255   0.08       0.205       0.0895        0.0395   
...   ...      ...      ...    ...         ...          ...           ...   
4172    F    0.565     0.45  0.165       0.887         0.37         0.239   
4173    M     0.59     0.44  0.135       0.966        0.439        0.2145   
4174    M      0.6    0.475  0.205       1.176       0.5255        0.2875   
4175    F    0.625    0.485   0.15      1.0945        0.531         0.261   
4176    M     0.71    0.555  0.195      1.9485       0.9455        0.3765   

     Peso_concha Anillos  
0           0.15      15  
1           0.07     

In [ ]:
# Verificamos si hay valores faltantes
print(df.isnull().sum())

# Si hay valores faltantes, eliminar filas con valores faltantes
df = df.dropna()


Sexo             0
Longitud         0
Diámetro         0
Altura           0
Peso_entero      0
Peso_manzana     0
Peso_vísceras    0
Peso_concha      0
Anillos          0
dtype: int64


In [ ]:
# Convertimos columnas numéricas a tipo numérico
df['Longitud'] = pd.to_numeric(df['Longitud'])
df['Diámetro'] = pd.to_numeric(df['Diámetro'])
df['Altura'] = pd.to_numeric(df['Altura'])
df['Peso_entero'] = pd.to_numeric(df['Peso_entero'])
df['Peso_manzana'] = pd.to_numeric(df['Peso_manzana'])
df['Peso_vísceras'] = pd.to_numeric(df['Peso_vísceras'])
df['Peso_concha'] = pd.to_numeric(df['Peso_concha'])
df['Anillos'] = pd.to_numeric(df['Anillos'])
df

Sexo  Longitud  Diámetro  Altura  Peso_entero  Peso_manzana  \
0       M     0.455     0.365   0.095       0.5140        0.2245   
1       M     0.350     0.265   0.090       0.2255        0.0995   
2       F     0.530     0.420   0.135       0.6770        0.2565   
3       M     0.440     0.365   0.125       0.5160        0.2155   
4       I     0.330     0.255   0.080       0.2050        0.0895   
...   ...       ...       ...     ...          ...           ...   
4172    F     0.565     0.450   0.165       0.8870        0.3700   
4173    M     0.590     0.440   0.135       0.9660        0.4390   
4174    M     0.600     0.475   0.205       1.1760        0.5255   
4175    F     0.625     0.485   0.150       1.0945        0.5310   
4176    M     0.710     0.555   0.195       1.9485        0.9455   

      Peso_vísceras  Peso_concha  Anillos  
0            0.1010       0.1500       15  
1            0.0485       0.0700        7  
2            0.1415       0.2100        9  
3            0.1140       0.1550       10  
4            0.0395       0.0550        7  
...             ...          ...      ...  
4172         0.2390       0.2490       11  
4173         0.2145       0.2605       10  
4174         0.2875       0.3080        9  
4175         0.2610       0.2960       10  
4176         0.3765       0.4950       12  

[4177 rows x 9 columns]

In [ ]:
# Codificamos la columna "Sexo" usando one-hot encoding
df = pd.get_dummies(df, columns=['Sexo'])
df

Longitud  Diámetro  Altura  Peso_entero  Peso_manzana  Peso_vísceras  \
0        0.455     0.365   0.095       0.5140        0.2245         0.1010   
1        0.350     0.265   0.090       0.2255        0.0995         0.0485   
2        0.530     0.420   0.135       0.6770        0.2565         0.1415   
3        0.440     0.365   0.125       0.5160        0.2155         0.1140   
4        0.330     0.255   0.080       0.2050        0.0895         0.0395   
...        ...       ...     ...          ...           ...            ...   
4172     0.565     0.450   0.165       0.8870        0.3700         0.2390   
4173     0.590     0.440   0.135       0.9660        0.4390         0.2145   
4174     0.600     0.475   0.205       1.1760        0.5255         0.2875   
4175     0.625     0.485   0.150       1.0945        0.5310         0.2610   
4176     0.710     0.555   0.195       1.9485        0.9455         0.3765   

      Peso_concha  Anillos  Sexo_F  Sexo_I  Sexo_M  
0          0.1500       15   False   False    True  
1          0.0700        7   False   False    True  
2          0.2100        9    True   False   False  
3          0.1550       10   False   False    True  
4          0.0550        7   False    True   False  
...           ...      ...     ...     ...     ...  
4172       0.2490       11    True   False   False  
4173       0.2605       10   False   False    True  
4174       0.3080        9   False   False    True  
4175       0.2960       10    True   False   False  
4176       0.4950       12   False   False    True  

[4177 rows x 11 columns]

In [ ]:
# Exploramos datos atípicos utilizando describe()
print(df.describe())

# Filtramos datos atípicos basados en umbrales específicos
# Por ejemplo, para eliminar filas con valores de Longitud extremadamente altos
df = df[df['Longitud'] < 1.0]  # Ajusta el umbral según sea necesario

# También podrías aplicar técnicas más avanzadas para identificar y manejar outliers


          Longitud     Diámetro       Altura  Peso_entero  Peso_manzana  \
count  4177.000000  4177.000000  4177.000000  4177.000000   4177.000000   
mean      0.523992     0.407881     0.139516     0.828742      0.359367   
std       0.120093     0.099240     0.041827     0.490389      0.221963   
min       0.075000     0.055000     0.000000     0.002000      0.001000   
25%       0.450000     0.350000     0.115000     0.441500      0.186000   
50%       0.545000     0.425000     0.140000     0.799500      0.336000   
75%       0.615000     0.480000     0.165000     1.153000      0.502000   
max       0.815000     0.650000     1.130000     2.825500      1.488000   

       Peso_vísceras  Peso_concha      Anillos  
count    4177.000000  4177.000000  4177.000000  
mean        0.180594     0.238831     9.933684  
std         0.109614     0.139203     3.224169  
min         0.000500     0.001500     1.000000  
25%         0.093500     0.130000     8.000000  
50%         0.171000     0.234

In [ ]:
from sklearn.preprocessing import StandardScaler

# Inicializaremos el escalador
scaler = StandardScaler()

# Normalizaremos las columnas numéricas
columnas_numericas = ['Longitud', 'Diámetro', 'Altura', 'Peso_entero', 'Peso_manzana', 'Peso_vísceras', 'Peso_concha', 'Anillos']
df[columnas_numericas] = scaler.fit_transform(df[columnas_numericas])


In [ ]:
# Calculamos el rango intercuartílico (IQR) para cada columna numérica
Q1 = df[columnas_numericas].quantile(0.25)
Q3 = df[columnas_numericas].quantile(0.75)
IQR = Q3 - Q1

# Filtramos filas que están fuera del rango IQR
df_sin_outliers = df[~((df[columnas_numericas] < (Q1 - 1.5 * IQR)) | (df[columnas_numericas] > (Q3 + 1.5 * IQR))).any(axis=1)]

# Imprimimos las filas que se eliminaron como outliers
filas_outliers = df[((df[columnas_numericas] < (Q1 - 1.5 * IQR)) | (df[columnas_numericas] > (Q3 + 1.5 * IQR))).any(axis=1)]
print("Filas eliminadas como outliers:")
print(filas_outliers)


Filas eliminadas como outliers:
      Longitud  Diámetro  Altura  Peso_entero  Peso_manzana  Peso_vísceras  \
6        0.530     0.415   0.150       0.7775        0.2370         0.1415   
7        0.545     0.425   0.125       0.7680        0.2940         0.1495   
9        0.550     0.440   0.150       0.8945        0.3145         0.1510   
32       0.665     0.525   0.165       1.3380        0.5515         0.3575   
33       0.680     0.550   0.175       1.7980        0.8150         0.3925   
...        ...       ...     ...          ...           ...            ...   
3993     0.720     0.600   0.235       2.2385        0.9840         0.4110   
3994     0.185     0.135   0.045       0.0320        0.0110         0.0065   
3996     0.315     0.230   0.000       0.1340        0.0575         0.0285   
4145     0.670     0.525   0.200       1.7405        0.6205         0.2970   
4148     0.770     0.605   0.175       2.0505        0.8005         0.5260   

      Peso_concha  Anillos  Sex

In [ ]:
df

Longitud  Diámetro  Altura  Peso_entero  Peso_manzana  Peso_vísceras  \
0        0.455     0.365   0.095       0.5140        0.2245         0.1010   
1        0.350     0.265   0.090       0.2255        0.0995         0.0485   
2        0.530     0.420   0.135       0.6770        0.2565         0.1415   
3        0.440     0.365   0.125       0.5160        0.2155         0.1140   
4        0.330     0.255   0.080       0.2050        0.0895         0.0395   
...        ...       ...     ...          ...           ...            ...   
4172     0.565     0.450   0.165       0.8870        0.3700         0.2390   
4173     0.590     0.440   0.135       0.9660        0.4390         0.2145   
4174     0.600     0.475   0.205       1.1760        0.5255         0.2875   
4175     0.625     0.485   0.150       1.0945        0.5310         0.2610   
4176     0.710     0.555   0.195       1.9485        0.9455         0.3765   

      Peso_concha  Anillos  Sexo_F  Sexo_I  Sexo_M  
0          0.1500       15   False   False    True  
1          0.0700        7   False   False    True  
2          0.2100        9    True   False   False  
3          0.1550       10   False   False    True  
4          0.0550        7   False    True   False  
...           ...      ...     ...     ...     ...  
4172       0.2490       11    True   False   False  
4173       0.2605       10   False   False    True  
4174       0.3080        9   False   False    True  
4175       0.2960       10    True   False   False  
4176       0.4950       12   False   False    True  

[4177 rows x 11 columns]

In [ ]:
from sklearn.model_selection import train_test_split

# División de los datos para la tarea de regresión
X_regresion = df.drop(['Anillos'], axis=1)  # Características (todas excepto 'Anillos')
y_regresion = df['Anillos']  # Variable objetivo ('Anillos')

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_regresion, y_regresion, test_size=0.2, random_state=42)

# División de los datos para la tarea de clasificación
X_clasificacion = df.drop(['Sexo_F', 'Sexo_I', 'Sexo_M'], axis=1)  # Características (todas excepto las columnas de sexo codificadas)
y_clasificacion = df[['Sexo_F', 'Sexo_I', 'Sexo_M']]  # Variables objetivo (columnas de sexo codificadas)

X_train_clas, X_test_clas, y_train_clas, y_test_clas = train_test_split(X_clasificacion, y_clasificacion, test_size=0.2, random_state=42)


In [ ]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

# Inicializaremos y entrenaremos el modelo KNN para regresión
knn_regresion = KNeighborsRegressor()
knn_regresion.fit(X_train_reg, y_train_reg)

# Predecimos en el conjunto de prueba
y_pred_reg = knn_regresion.predict(X_test_reg)

# Calculamos el error cuadrático medio (MSE)
mse_reg = mean_squared_error(y_test_reg, y_pred_reg)
print("Error cuadrático medio (MSE) para regresión:", mse_reg)


Error cuadrático medio (MSE) para regresión: 5.131483253588517


El desempeño del modelo KNN para regresión en la tarea de predecir la edad del abulón fue evaluado utilizando el Error cuadrático medio (MSE). El MSE calculado fue de aproximadamente 5.13. Esto indica que, en promedio, las predicciones del modelo tienen un error cuadrático medio de 5.13 unidades en relación con los valores reales de la edad del abulón. En general, un MSE bajo sugiere que el modelo puede realizar buenas predicciones, aunque el contexto específico de la aplicación determinará si este nivel de precisión es adecuado para el propósito previsto.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

# Inicializaremos y entrenaremos el modelo KNN para clasificación
knn_clasificacion = KNeighborsClassifier()
knn_clasificacion.fit(X_train_clas, y_train_clas)

# Predeciremos en el conjunto de prueba
y_pred_clas = knn_clasificacion.predict(X_test_clas)

# Calcularemos la precisión del modelo
accuracy_clas = accuracy_score(y_test_clas, y_pred_clas)
print("Precisión del modelo KNN para clasificación:", accuracy_clas)

# Mostraremos el reporte de clasificación
print("\nReporte de clasificación:")
print(classification_report(y_test_clas, y_pred_clas))


Precisión del modelo KNN para clasificación: 0.49760765550239233

Reporte de clasificación:
              precision    recall  f1-score   support

           0       0.45      0.35      0.39       262
           1       0.77      0.70      0.73       267
           2       0.50      0.45      0.48       307

   micro avg       0.58      0.50      0.53       836
   macro avg       0.57      0.50      0.53       836
weighted avg       0.57      0.50      0.53       836
 samples avg       0.50      0.50      0.50       836



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


El desempeño del modelo KNN para clasificación se evaluó utilizando la precisión del modelo y el reporte de clasificación. La precisión del modelo fue aproximadamente del 49.76%. Esto indica que el modelo clasificó correctamente alrededor del 49.76% de las muestras en el conjunto de prueba.

Al analizar el reporte de clasificación, observamos que las métricas de precisión, recall y f1-score varían para cada clase (0, 1 y 2), lo que indica cómo el modelo se desempeñó en la clasificación de cada clase en comparación con las otras clases. Además, la métrica de precisión promedio ponderada (weighted avg) y la métrica de precisión promedio macro (macro avg) proporcionan una visión general del desempeño general del modelo en todas las clases. En este caso, tanto la precisión promedio ponderada como la precisión promedio macro son aproximadamente del 57%, lo que sugiere que el modelo tiene un desempeño moderado en la clasificación de las clases.

La métrica samples avg proporciona la precisión promedio sobre las muestras reales, que es del 50%, lo que indica que el modelo tiene un rendimiento medio en la clasificación general de las muestras.

¿Podrías haber utilizado la regresión lineal y la regresión KNN para resolver el problema de regresión?

Sí, tanto la regresión lineal como la regresión KNN son técnicas que pueden aplicarse para resolver problemas de regresión. Sin embargo, es importante considerar las características del conjunto de datos y las suposiciones de cada modelo al elegir la técnica más adecuada.

Regresión lineal: Es una técnica que asume una relación lineal entre las características y la variable objetivo. Se puede utilizar cuando hay una relación lineal entre las características y la variable objetivo, y cuando se cumplen las suposiciones de la regresión lineal, como la normalidad de los residuos y la independencia de estos.
Regresión KNN (K-Nearest Neighbors): Es un algoritmo de aprendizaje supervisado no paramétrico que se utiliza tanto para problemas de regresión como de clasificación. Funciona encontrando los "k" puntos de datos más cercanos en el espacio de características y prediciendo el valor de la variable objetivo basado en la media (en el caso de regresión) de estos puntos. Puede ser útil cuando no se conoce la forma de la relación entre las características y la variable objetivo, y cuando no se cumplen las suposiciones de la regresión lineal.
En resumen, tanto la regresión lineal como la regresión KNN podrían haber sido opciones viables para resolver el problema de regresión, y la elección entre ellas dependería de la naturaleza del conjunto de datos y de si se cumplen las suposiciones de la regresión lineal.



¿Podrías haber utilizado la regresión lineal para el problema de clasificación?


No, la regresión lineal no es una técnica adecuada para resolver problemas de clasificación. La regresión lineal se utiliza para predecir valores continuos, mientras que los problemas de clasificación implican predecir clases o categorías discretas. La regresión lineal no está diseñada para manejar la naturaleza discreta de las variables objetivo en problemas de clasificación.

Para los problemas de clasificación, se suelen utilizar técnicas específicas como la regresión logística, los árboles de decisión, las máquinas de vectores de soporte (SVM) o algoritmos de vecinos más cercanos (KNN) para clasificación, entre otros. Estos algoritmos están diseñados para manejar la naturaleza discreta de las variables objetivo y son más apropiados para problemas de clasificación.